cmd/geth/snapshot.go

```diff
@@ -0,0 +1,413 @@
+					utils.RinkebyFlag,
+					utils.GoerliFlag,
+					utils.LegacyTestnetFlag,
+				},
+				Description: `
+geth snapshot verify-state <state-root>
+will traverse the whole accounts and storages set based on the specified
+snapshot and recalculate the root hash of state for verification.
+In other words, this command does the snapshot to trie conversion.
+`,
+			},
+			{
+				Name:      "traverse-state",
+				Usage:     "Traverse the state with given root hash for verification",
+				ArgsUsage: "<root>",
+				Action:    utils.MigrateFlags(traverseState),
+				Category:  "MISCELLANEOUS COMMANDS",
+				Flags: []cli.Flag{
+					utils.DataDirFlag,
+					utils.RopstenFlag,
+					utils.RinkebyFlag,
+					utils.GoerliFlag,
+					utils.LegacyTestnetFlag,
+				},
+				Description: `
+geth snapshot traverse-state <state-root>
+will traverse the whole state from the given state root and will abort if any
+referenced trie node or contract code is missing. This command can be used for
+state integrity verification. The default checking target is the HEAD state.
+
+It's also usable without snapshot enabled.
+`,
+			},
+			{
+				Name:      "traverse-rawstate",
+				Usage:     "Traverse the state with given root hash for verification",
+				ArgsUsage: "<root>",
+				Action:    utils.MigrateFlags(traverseRawState),
+				Category:  "MISCELLANEOUS COMMANDS",
+				Flags: []cli.Flag{
+					utils.DataDirFlag,
+					utils.RopstenFlag,
+					utils.RinkebyFlag,
+					utils.GoerliFlag,
+					utils.LegacyTestnetFlag,
+				},
+				Description: `
+geth snapshot traverse-rawstate <state-root>
+will traverse the whole state from the given root and will abort if any referenced
+trie node or contract code is missing. This command can be used for state integrity
+verification. The default checking target is the HEAD state. It's basically identical
+to traverse-state, but the check granularity is smaller. 
+
+It's also usable without snapshot enabled.
+`,
+			},
+		},
+	}
+)
+
+func pruneState(ctx *cli.Context) error {
+	glogger := log.NewGlogHandler(log.StreamHandler(os.Stderr, log.TerminalFormat(true)))
+	glogger.Verbosity(log.LvlInfo)
+	log.Root().SetHandler(glogger)
+
+	stack, _ := makeConfigNode(ctx)
+	defer stack.Close()
+
+	chain, chaindb := utils.MakeChain(ctx, stack, true)
+	defer chaindb.Close()
+
+	trieCache := eth.DefaultConfig.TrieCleanCacheJournal
+	if ctx.GlobalIsSet(utils.CacheTrieJournalFlag.Name) {
+		trieCache = ctx.GlobalString(utils.CacheTrieJournalFlag.Name)
+		log.Info("Customized trie clean cache specified", "path", stack.ResolvePath(trieCache))
+	}
+	pruner, err := pruner.NewPruner(chaindb, chain.CurrentBlock().Header(), stack.ResolvePath(""), trieCache, ctx.GlobalUint64(utils.BloomFilterSizeFlag.Name))
+	if err != nil {
+		utils.Fatalf("Failed to open snapshot tree %v", err)
+	}
+	if ctx.NArg() > 1 {
+		utils.Fatalf("too many arguments given")
+	}
+	var targetRoot common.Hash
+	if ctx.NArg() == 1 {
+		targetRoot = common.HexToHash(ctx.Args()[0])
```

Comment:
`common.HexToHash` is useful internally if you know that the string is valid, but you should really not call it on untrusted user input because it just blindly converts bytes to a hash.

I think we should use `hexutil.Decode` to turn the input into a string (and check any errors) and then if the input is 32 bytes, then we can blindly convert via `common.BytesToHash`.

Maybe we could even make a `common.NewHash(hex string) (common.Hash, error)` method to make this simpler in general and less error prone for future code. CC @fjl ?

Reply:
You can also use UnmarshalText to parse it, like

```
var h common.Hash
if err := h.UnmarshalText(input); err != nil {
    ...
}
```

Reply:
Even better, less API, thx

Reply:
Sorry, UnmarshalText actually requires its parameter to be `[]byte`. You can still use it, just requires a conversion.

Reply:
Fixed